Tests/Engine/Results/BacktestingResultHandlerTests.cs

```diff
@@ -0,0 +1,391 @@
+ * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+*/
+
+using Deedle;
+using NUnit.Framework;
+using QuantConnect.Algorithm.CSharp;
+using QuantConnect.Lean.Engine.Results;
+using QuantConnect.Report;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace QuantConnect.Tests.Engine.Results
+{
+    [TestFixture]
+    public class BacktestingResultHandlerTests
+    {
+        private Dictionary<string, BacktestingResultHandler> _resultsCache = new Dictionary<string, BacktestingResultHandler>();
+
+        public BacktestingResultHandler GetResults(string algorithm, DateTime algoStart, DateTime algoEnd)
+        {
+            BacktestingResultHandler backtestingResult;
+            if (!_resultsCache.TryGetValue(algorithm, out backtestingResult))
+            {
+                var parameter = new RegressionTests.AlgorithmStatisticsTestParameters(algorithm,
+                   new Dictionary<string, string>(),
+                   Language.CSharp,
+                   AlgorithmStatus.Completed);
+
+
+                // The AlgorithmRunner uses the `RegressionResultHandler` but doesn't do any sampling.
+                // It defaults to the behavior of the `BacktestingResultHandler` class in `results.ProcessSynchronousEvents()`
+                AlgorithmRunner.RunLocalBacktest(parameter.Algorithm,
+                    parameter.Statistics,
+                    parameter.AlphaStatistics,
+                    parameter.Language,
+                    parameter.ExpectedFinalStatus,
+                    startDate: algoStart,
+                    endDate: algoEnd,
+                    storeResult: true);
+
+                _resultsCache[algorithm] = AlgorithmRunner.AlgorithmResults[Language.CSharp][algorithm];
+                AlgorithmRunner.AlgorithmResults[Language.CSharp].Remove(algorithm);
+            }
+
+            return _resultsCache[algorithm];
+        }
+
+        [TestCase(nameof(BasicTemplateAlgorithm))]
+        [TestCase(nameof(BasicTemplateDailyAlgorithm))]
+        [TestCase(nameof(ResolutionSwitchingAlgorithm))]
+        public void SamplesNotMisalignedRelative(string algorithm)
+        {
+            var backtestResults = GetResults(algorithm, new DateTime(2013, 10, 7), new DateTime(2013, 10, 11));
+            var benchmarkSeries = backtestResults.Charts["Benchmark"].Series["Benchmark"];
+            var equitySeries = backtestResults.Charts["Strategy Equity"].Series["Equity"];
+            var performanceSeries = backtestResults.Charts["Strategy Equity"].Series["Daily Performance"];
```

Comment:
I think we are also missing tests that go through the code paths in `StatisticsBuilder`, where part of the changes took place? Maybe simpler unit tests, just for the `StatisticsBuilder`, where for example we can assert it throws if given "misaligned values", it resamples daily